### PR TITLE
Fix keySet(), entrySet(), values() in IdentityLinkedHashMap

### PR DESCRIPTION
**Pull request by findepi**
`IdentityLinkedHashMap` is an identity-based hash map (with stable iteration order).

But `newSetFromMap(new IdentityLinkedHashMap())` is _not_ an identity-based set. Since such sets are used extensively during analysis, they _must_ be correct.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/util/maps/IdentityLinkedHashMap.java [hunk 1]
```diff
@@ -142,4 +133,116 @@ public int size()
             }
         };
     }
+
+    private class KeySet
+            extends SetView<K>
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/util/maps/IdentityLinkedHashMap.java]: Dear Reviewer, 

design decision here. I could extend `AbstractSet` and override only what needed (this is what `IdentityHashMap.EntrySet` does), but I find this wrong thing to do as long as I'm not part-time JDK collections dev. `AbstractSet` might be OK for well-behaved sets, which we're not, so it's not super obvious what needs to be overridden today and how that would work tomorrow.

Thus I traded code simplicity for code conciseness.

File: presto-main/src/main/java/com/facebook/presto/util/maps/IdentityLinkedHashMap.java [hunk 2]
```diff
@@ -142,4 +133,116 @@ public int size()
             }
         };
     }
+
+    private class KeySet
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/util/maps/IdentityLinkedHashMap.java]: What prompted this change? When this class was first added I remember thinking it looked odd but it didn't seem broken, just limited/potentially inefficient (mutation methods in AbstractSet throw, everything else delegates to iterator()).

File: presto-main/src/main/java/com/facebook/presto/util/maps/IdentityLinkedHashMap.java [hunk 3]
```diff
@@ -142,4 +133,116 @@ public int size()
...
+        }
+
+        @Override
+        public boolean remove(Object o)
+        {
+            return delegate.entrySet().remove(equivalence.wrap(o));
+        }
+
+        @Override
+        public boolean retainAll(Collection<?> c)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/util/maps/IdentityLinkedHashMap.java]: Don't abbreviate. Call the argument "other".

File: presto-main/src/main/java/com/facebook/presto/util/maps/IdentityLinkedHashMap.java [hunk 4]
```diff
@@ -142,4 +133,116 @@ public int size()
...
+        }
+
+        @Override
+        public Iterator<K> iterator()
+        {
+            return Iterators.transform(delegate.keySet().iterator(), Equivalence.Wrapper::get);
+        }
+
+        @Override
+        public boolean remove(Object o)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/util/maps/IdentityLinkedHashMap.java]: Don't abbreviate. Call the argument "value", "item", etc.

File: presto-main/src/main/java/com/facebook/presto/util/maps/IdentityLinkedHashMap.java [hunk 5]
```diff
@@ -142,4 +133,116 @@ public int size()
             }
         };
     }
+
+    private class KeySet
+            extends SetView<K>
+    {
+        @Override
+        public boolean contains(Object o)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/util/maps/IdentityLinkedHashMap.java]: Don't abbreviate. Call the argument "value", "item", etc.

File: presto-main/src/main/java/com/facebook/presto/util/maps/IdentityLinkedHashMap.java (hunk 2 above)
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/util/maps/IdentityLinkedHashMap.java]: In #7354 i created `IndetityLinkedHashMap columnReferences` and wrote `columnReferences.keySet().contains(..)` -- equality comparison was used.

(Of course, I should write `columnReferences.containsKey(..)` but that's not the point)

File: presto-main/src/main/java/com/facebook/presto/util/maps/IdentityLinkedHashMap.java (hunk 5 above)
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/util/maps/IdentityLinkedHashMap.java]: Eh... I didn't abbreviate. These are "official" names in the implemented interface. I will change though.

**Comment by findepi** [comment]
Comments applied (as `fixup` commits). 
Thanks @martint  for the review!

File: presto-main/src/main/java/com/facebook/presto/util/maps/IdentityLinkedHashMap.java [hunk 6]
```diff
@@ -165,6 +148,46 @@ public boolean retainAll(Collection<?> other)
...
+        {
+            if (!(o instanceof Entry)) {
+                return false;
+            }
+            Entry<?, ?> entry = (Entry<?, ?>) o;
+            Equivalence.Wrapper<?> wrappedKey = equivalence.wrap(entry.getKey());
+            return delegate.get(wrappedKey) == entry.getValue()
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/util/maps/IdentityLinkedHashMap.java]: containsValue() & values() are equality based (not identity based)

File: presto-main/src/main/java/com/facebook/presto/util/maps/IdentityLinkedHashMap.java (hunk 2 above)
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/util/maps/IdentityLinkedHashMap.java]: Well, then, unless we need all this code, I would prefer not adding it. It's just more stuff that can go wrong, that we have to maintain, etc.

File: presto-main/src/main/java/com/facebook/presto/util/maps/IdentityLinkedHashMap.java (hunk 6 above)
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/util/maps/IdentityLinkedHashMap.java]: Both fixed, thanks for catching.

File: presto-main/src/main/java/com/facebook/presto/util/maps/IdentityLinkedHashMap.java (hunk 2 above)
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/util/maps/IdentityLinkedHashMap.java]: But the code that "can go wrong" is already there. At the minimum, all broken methods should throw.

- keySet
- entrySet
- values
- containsValue

Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/util/maps/IdentityLinkedHashMap.java]: How are those broken, other than returning a read-only or inefficient collection? But I agree, if we don't need them, we should probably just make them throw.

Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/util/maps/IdentityLinkedHashMap.java]: `new IdentityHashMap().keySet().contains(...)` compares (expectedly) by identity.

`new IdentityLinkedHashMap().keySet().contains(...)` compares by equality.

Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/util/maps/IdentityLinkedHashMap.java]: I looked into #7354 and I'm using `unmodifiableSet(map.keySet())` there. If it's ok I'm using the non-linked version, then I'm ok with just throwing in every currently-broken method in ILHM.

**Comment by findepi** [comment]
Superseded by #7785

**Pull request closed without merging**